### PR TITLE
planner: add MergeAdjacentWindow rule for cascades

### PR DESCRIPTION
**Pull request by An-DJ**
### What problem does this PR solve?
Issue number: #13709

Problem Summary: 
This PR adds Transformation rule MergeAdjacentWindow, which is a part of MergeAdjacentPlans in cascades planner.

### What is changed and how it works?

What's Changed:
Add new cascades rule `MergeAdjacentWindow`

How it Works:
Use `MergeAdjacentWindow` to merge adjacent Window.

### Related changes

- none

### Check List 

Tests 

- Unit test

Side effects

- none

### Release note 

- No release note

### CODE REVIEW

**Comment by An-DJ** [comment]
/cc @francis0407 PTAL

I fix the `unstable problem` in PR #22853 which was mentioned in #14880 .  This rule can be stable in test.

File: planner/cascades/testdata/transformation_rules_suite_out.json
Comment (francis0407) [inline planner/cascades/testdata/transformation_rules_suite_out.json:2638]: I think the two windows should not be merged in this test?

File: planner/cascades/transformation_rules.go
Comment (francis0407) [inline planner/cascades/transformation_rules.go:2580]: You should use the schema of the `Group` instead of the `LogicalPlan`.

File: planner/cascades/testdata/transformation_rules_suite_out.json
Comment (An-DJ) [reply planner/cascades/testdata/transformation_rules_suite_out.json:2638]: The SQL:
```
select a, b, rank() over (partition by a), sum(bb) over (partition by a) as 'sum_bb', c from (select a, b, c, max(b) over (partition by a) as 'bb' from t) as tt
```
Before cascades:
```
Group#0 Schema:[test.t.a,test.t.b,Column#21,Column#22,test.t.c]
    Projection_12 input:[Group#1], test.t.a, test.t.b, Column#18, Column#17, test.t.c
Group#1 Schema:[test.t.a,test.t.b,Column#18,Column#17,test.t.c]
    Projection_11 input:[Group#2], test.t.a, test.t.b, Column#18, Column#17, test.t.c
Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#17,Column#18]
    Window_10 input:[Group#3]
Group#3 Schema:[test.t.a,test.t.b,test.t.c,Column#17]
    Projection_9 input:[Group#4], test.t.a, test.t.b, test.t.c, Column#17
Group#4 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17]
    Window_8 input:[Group#5]
Group#5 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_7 input:[Group#6], test.t.a, test.t.b, test.t.c, Column#14
Group#6 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_6 input:[Group#7], test.t.a, test.t.b, test.t.c, Column#14
Group#7 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_5 input:[Group#8], test.t.a, test.t.b, test.t.c, Column#14
Group#8 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Window_4 input:[Group#9]
Group#9 Schema:[test.t.a,test.t.b,test.t.c]
    Projection_3 input:[Group#10], test.t.a, test.t.b, test.t.c
Group#10 Schema:[test.t.a,test.t.b,test.t.c]
    Projection_2 input:[Group#11], test.t.a, test.t.b, test.t.c
Group#11 Schema:[test.t.a,test.t.b,test.t.c]
    DataSource_1 table:t
```

After cascades (rules `EliminateProjection` and `MergeAdjacentProjection`):
```
Group#0 Schema:[test.t.a,test.t.b,Column#21,Column#22,test.t.c]
     Projection_13 input:[Group#1], test.t.a, test.t.b, Column#18, Column#17, test.t.c 
Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#17,Column#18]
     Window_10 input:[Group#2] 
Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#17]
     Projection_9 input:[Group#3], test.t.a, test.t.b, test.t.c, Column#17 
Group#3 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17]
     Window_8 input:[Group#4] 
Group#4 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
     Window_4 input:[Group#5] 
Group#5 Schema:[test.t.a,test.t.b,test.t.c]
     DataSource_1 table:t]
```

I think the Group#2 Projection should be eliminated by  `EliminateProjection` or other rules. 

`MergeAdjacentWindow` can only merge adjacent window groups so that the two window function cannot be merged.  Isn't that true?

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go:2580]: Is the schema in `Group` the same as in `LogicalPlan` 😄?

Comment (francis0407) [reply planner/cascades/transformation_rules.go:2580]: No, they are not equivalent. 

In cascades planner, the schema of `LogicalPlan` should never be used. We don't set the schema of new created `LogicalPlan`s. Every time we want to set or get the schema, we use the schema of `Group` instead.

File: planner/cascades/testdata/transformation_rules_suite_out.json
Comment (francis0407) [reply planner/cascades/testdata/transformation_rules_suite_out.json:2638]: Emmm, you may misunderstand me. 

The SQL:
```
select a, b, sum(bb) over (partition by a) as 'sum_bb', c, rank() over (partition by a) 
from (
    select a, b, c, max(b) over (partition by a) as 'bb' 
    from t
) as tt
```

I think this SQL should generate a plan with two Windows.  Because the outer one reference `bb`, which is the result of the inner one. So the two windows can not be merged.

But the test result only contains one:
```


"Result": [
          "Group#0 Schema:[test.t.a,test.t.b,Column#21,test.t.c,Column#23]",
          "    Projection_15 input:[Group#1], test.t.a, test.t.b, Column#18, test.t.c, Column#17",
          "Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17,Column#18]",
          "    Window_14 input:[Group#2]",          "Group#2 Schema:[test.t.a,test.t.b,test.t.c]",
          "    DataSource_1 table:t"
```

Comment (An-DJ) [reply planner/cascades/testdata/transformation_rules_suite_out.json:2638]: Thanks for your guidance! This bug is produced by the abuse of  `LogicalPlan` schema. 😸  I fix it in new commit just now.

**Comment by tisonkun** [comment]
/uncc @tiancaiamao @time-and-fate @winoros 

Hi @An-DJ ! Thanks for your contribution! @eurekaka is working on the cascaded planner. I will try to ping him in privacy. Other people in-cc do not focus on this domain so they may not provide effective reviews. Please avoid ping a large range of people which brings noise.

Also according to well recognized open source principle:

> Be patient if your patch is not applied as fast as you'd like (open source developers are all volunteers, often doing the development in their spare time) or a developer asks you to make changes to the patch. If you do not receive any feedback in a reasonable amount of time (say, a week or two), feel free to post a follow-up topic on our [developer discussion forum](https://internals.tidb.io/).

**Comment by An-DJ** [comment]
> /uncc @tiancaiamao @time-and-fate @winoros
> 
> Hi @An-DJ ! Thanks for your contribution! @eurekaka is working on the cascaded planner. I will try to ping him in privacy. Other people in-cc do not focus on this domain so they may not provide effective reviews. Please avoid ping a large range of people which brings noise.
> 
> Also according to well recognized open source principle:
> 
> > Be patient if your patch is not applied as fast as you'd like (open source developers are all volunteers, often doing the development in their spare time) or a developer asks you to make changes to the patch. If you do not receive any feedback in a reasonable amount of time (say, a week or two), feel free to post a follow-up topic on our [developer discussion forum](https://internals.tidb.io/).

Thanks for your nice reply!

Sorry to bother so many reviewers 😿 I'm not sure which Reviewers focus on the Planner domain, so I remind it to everyone who followed the `planner` label. I will wait patiently~

Thanks again.

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:2564]: Should we check the `Frame.End` are the same too?

**Comment by time-and-fate** [comment]
@An-DJ Currently our team doesn't focus on the development of the cascades planner so PRs for the cascades planner will be very slow to be reviewed. Sorry about that.
I'll take some time to review this PR maybe next week.

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:2622]: `newWindowFuncs := make([]*aggregation.WindowFuncDesc, 0, len(curWinPlan.WindowFuncDescs)+len(nextWinPlan.WindowFuncDescs))`

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go:2567]: Thank you for your correction. I have fixed it in [`commit 
42eb21c`](https://github.com/pingcap/tidb/pull/24556/commits/42eb21c36125a6190631a61f82377f66a81004b9).

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:2545]: The `partition_clause` has this syntax: `PARTITION BY expr [, expr] ...`, so we don't need to check `Desc` of `PartitionBy`

Is the result of the `partition by a, b` same as `partition by b, a`? 
Should we merge them in this case?

Comment (An-DJ) [reply planner/cascades/transformation_rules.go:2545]: > The `partition_clause` has this syntax: `PARTITION BY expr [, expr] ...`, so IMO, we don't need to check `Desc` of `PartitionBy`
> 

> Is the result of the `partition by a, b` same as `partition by b, a`?
> Should we merge them in this case?

@mmyj Maybe you are right. I hold that the columns in partition clause cannot affect the result of window, but the implementation of partititon clause (`property.SortItem`) makes me confused. 

By the way, why parititon clause is implemented by `SortItem`?

Comment (mmyj) [reply planner/cascades/transformation_rules.go:2545]: I'm confused also.

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:134]: Why this rule set into the `TiKVLayerOptimizationBatch`?

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:2576]: Maybe we could implement an `Equal` function for the `FrameBound` structure,  it will simplify the code here.

File: planner/cascades/transformation_rules.go
Comment (mmyj) [inline planner/cascades/transformation_rules.go:2589]: IMO, we don't need to loop twice and reuse the result here. How about following code?
```
	nextWindowCols := make(map[int64]struct{})
	nextWindowChildrenExistedCols := make(map[int64]struct{})
	nextSchema := nextGroupExpr.Group.Prop.Schema
	nextGroupChildren := nextGroupExpr.Children
	for _, ngc := range nextGroupChildren {
		for _, c := range ngc.Prop.Schema.Columns {
			nextWindowChildrenExistedCols[c.UniqueID] = struct{}{}
		}
	}
	for _, nc := range nextSchema.Columns {
		nextWindowCols[nc.UniqueID] = struct{}{}
	}
	for _, funDesc := range curWinPlan.WindowFuncDescs {
		for _, arg := range funDesc.Args {
			cols := expression.ExtractColumns(arg)
			for _, c := range cols {
				_, ok1 := nextWindowCols[c.UniqueID]
				_, ok2 := nextWindowChildrenExistedCols[c.UniqueID]
				if ok1 && !ok2 {
					return false
				}
			}
		}
	}
```

Comment (mmyj) [reply planner/cascades/transformation_rules.go:2589]: I'm not sure that using `UniqueID` here is correct.

Comment (An-DJ) [reply planner/cascades/transformation_rules.go:2589]: > I'm not sure that using `UniqueID` here is correct.

Well example. I think using `UniqueID` is correct here because the column compare function `Equal` is implemented by `UniqueID`, too. I will modify that soon.

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go:134]: > Why this rule set into the `TiKVLayerOptimizationBatch`?

My fault. I will adjust it to `TiDBLayerOptimizationBatch`.

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go]: > I'm confused also.

@mmyj I remove the order-consistent-force of `PartitionBy` in [`newer commit`](https://github.com/pingcap/tidb/pull/24556/commits/071f638ba9a88c5b87f105b7b5693276defaf37a) just now.

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go]: @mmyj I have adjusted the rule area to `TiDBLayerOptimizationBatch` and `Equal` judge codes in [newer commit](https://github.com/pingcap/tidb/pull/24556/commits/071f638ba9a88c5b87f105b7b5693276defaf37a) just now.

**Comment by mmyj** [comment]
@An-DJ It's LGTM, but I'm not a reviewer of the sig-planner. This PR need to be wait to others to review.

File: planner/cascades/transformation_rules.go
Comment (time-and-fate) [inline planner/cascades/transformation_rules.go:2565]: It seems that the `!ok2` check is enough?

**Review by time-and-fate — COMMENTED** [review]
others LGTM.

File: planner/cascades/transformation_rules.go
Comment (An-DJ) [reply planner/cascades/transformation_rules.go:2565]: > It seems that the `!ok2` check is enough?

@time-and-fate Maybe you are right. I considered the case `SELECT COUNT(1) OVER (...) FROM (SELECT COUNT(1) OVER (...) FROM T) AS TT` before, but I realize now that arg `1` can be ignored by `ExtractColumns()`. Maybe all columns in currentWindow arguments can be found in the nextWindowSchema.

Comment (An-DJ) [reply planner/cascades/transformation_rules.go]: @time-and-fate Done. (https://github.com/pingcap/tidb/pull/24556/commits/4b0ec3fb502d4e1dbe042d8a379fe09937d354b8)

**Review by time-and-fate — APPROVED** [review]
LGTM. Thanks for your contribution :)

**Comment by tisonkun** [comment]
@time-and-fate thanks for your review! We are waiting for another approval to merge this PR. Can you cc other potential reviewers?

**Comment by time-and-fate** [comment]
> @time-and-fate thanks for your review! We are waiting for another approval to merge this PR. Can you cc other potential reviewers?

OK.
@francis0407 Could you take a look again?

File: planner/cascades/testdata/transformation_rules_suite_out.json
Comment (francis0407) [inline planner/cascades/testdata/transformation_rules_suite_out.json:2698]: It's difficult to figure out which window function this plan is corresponding to from this result.

I think we'd better add some detailed information about the window operator in the explained result, e.g. the window function name and the partition columns. 

Perhaps we can fix this in another PR.

**Review by francis0407 — APPROVED** [review]
LGTM

**Comment by jackwener** [comment]
@eurekaka PTAL

**Comment by jackwener** [comment]
There are some error in the test. You can see the details in the `Details`.

**Comment by jackwener** [comment]
cc @francis0407 
There are two `LGTM`. Now it need a committer merge this PR if there isn't problem.

**Pull request merged**
